File: python/src/resp/serializer.py
class RespSerializer:
  CRLF = '\r\n'
  
  @classmethod
  def serialize(cls, data: any) -> bytes:
    # Consider bulk string
    # Ignore simple strings
    result:str = ''
    if data is None:
      result = f'$-1{cls.CRLF}'
    elif type(data) == bytes:
      result = f'${len(data)}{cls.CRLF}{data.decode('utf-8')}{cls.CRLF}'
    elif type(data) == int:
      result = f':{data}{cls.CRLF}'
    elif type(data) == list:
      header: str = f'*{len(data)}{cls.CRLF}'.encode('UTF-8')
      payload: str = b''.join([cls.serialize(datum) for datum in data])
      return b''.join([header, payload])
    elif type(data) == Exception:
      result = f'-{data}{cls.CRLF}'
    else:
      raise Exception(f"RESP Serialization failed. Invalid class: {type(data)}")
    
    return result.encode('UTF-8')

File: ruby/lib/redis/resp/serializer.rb
module Redis
  module Resp
    class Serializer
      CRLF = "\r\n"
       
      def self.serialize(data)
        case data
        when NilClass
          "$-1#{CRLF}"
        when Symbol # Simple string
          "+#{data}#{CRLF}"
        when String # Bulk Strings
          "$#{data.bytesize}#{CRLF}#{data}#{CRLF}"
        when Integer
          ":#{data}#{CRLF}"
        when Array
          header = "*#{data.size}#{CRLF}"
          payload = data.map { |datum| serialize(datum) }.join
          "#{header}#{payload}"
        when StandardError
          "-#{data}#{CRLF}"
        else
          raise "RESP Serialization failed. Invalid class: #{data.class}"
        end
      end
    end
  end
end